File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        ArrayList<Integer> result = new ArrayList<>();
        
        HashMap<String, Integer> typeMap = new HashMap<>();
        
        // HashMap에 값 넣기
        for(int i = 0; i < terms.length; i++){
            String[] temp = terms[i].split(" "); 
            String type = temp[0];
            int term = Integer.parseInt(temp[1]); 
    
            typeMap.put(type, term);
        }
        
        // 오늘 날짜 변환
        String[] todayParts = today.split("\\.");
        int todayTotal = Integer.parseInt(todayParts[0]) * 12 * 28 + 
                         Integer.parseInt(todayParts[1]) * 28 + 
                         Integer.parseInt(todayParts[2]);
        
        // privacies 날짜 변환
        for(int i = 0; i < privacies.length; i++){
            String[] p = privacies[i].split(" ");
            String[] dateParts = p[0].split("\\.");
            int pTotal = Integer.parseInt(dateParts[0]) * 12 * 28 + 
                         Integer.parseInt(dateParts[1]) * 28 + 
                         Integer.parseInt(dateParts[2]);
            
            // 파기 시점 = 수집일 + (약관 개월 * 28)
            int expireDate = pTotal + (typeMap.get(p[1]) * 28);
            
            if(expireDate <= todayTotal) {
                result.add(i + 1);
            }
        }
        
        // 5. ArrayList를 int[]로 변환
        int[] answer = new int[result.size()];
        for(int i = 0; i < result.size(); i++) {
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    
    ty, tm, td = map(int, today.split("."))
    today_total = ty * 12 * 28 + tm * 28 + td
    
    type_dict = {s.split()[0]: int(s.split()[1]) * 28 for s in terms}
    
    answer = []
    
    for i in range(len(privacies)):
        day, p_type = privacies[i].split()
        py, pm, pd = map(int, day.split("."))
        
        # 개인정보 수집 일자를 총 일수로 변환
        privacy_total = py * 12 * 28 + pm * 28 + pd
        
        # 파기되어야 할 날짜 계산
        expire_date = privacy_total + type_dict[p_type]
        
        # 오늘 날짜와 비교
        # 파기일이 오늘보다 작거나 같다면 파기 대상
        if expire_date <= today_total:
            answer.append(i + 1) # i는 0부터 시작하므로 +1
            
    return answer